File: backend/app/schemas/admin/strategy_config.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
# @author  : Liu Lijun
# @date    : 2026-06-11
# @description: Strategy Config Pydantic Schemas

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# StrategyConfig schemas
# ---------------------------------------------------------------------------

class StrategyConfigBase(BaseModel):
    """Shared readable fields."""
    prompt_language: Optional[str] = None

    # Component switches
    enable_query_rewrite: bool = True
    enable_query_expansion: bool = False
    enable_query_hyde: bool = False
    enable_vector: bool = True
    enable_bm25: bool = True
    enable_reranker: bool = True
    enable_small_to_big: bool = True
    require_citation: bool = True

    # Search parameters
    query_expansion_num: int = 3
    max_transform_queries: int = 5
    vector_top_k: int = 30
    bm25_top_k: int = 30
    rrf_mode: str = "rrf"
    rrf_k: int = 60
    rrf_top_k: int = 20
    vector_weight: float = 0.6
    reranking_mode: str = "hybrid"
    rerank_top_k: int = 10
    final_top_k: int = 3

    # Threshold parameters
    vector_score_threshold: float = 0.5
    bm25_score_threshold: float = 0.1

    # Confidence
    confidence_high_score_threshold: float = 0.7
    confidence_min_high_conf_count: int = 1
    confidence_low_score_threshold: float = 0.5

    # Extra
    extra_config: Optional[dict[str, Any]] = None


class StrategyConfigCreate(StrategyConfigBase):
    """Fields required when creating a new config."""
    config_id: str = Field(..., max_length=100)
    kb_id: int
    version: int
    created_by: Optional[str] = None


class StrategyConfigUpdate(StrategyConfigBase):
    """All fields optional for partial update."""
    prompt_language: Optional[str] = None
    enable_query_rewrite: Optional[bool] = None
    enable_query_expansion: Optional[bool] = None
    enable_query_hyde: Optional[bool] = None
    enable_vector: Optional[bool] = None
    enable_bm25: Optional[bool] = None
    enable_reranker: Optional[bool] = None
    enable_small_to_big: Optional[bool] = None
    require_citation: Optional[bool] = None
    query_expansion_num: Optional[int] = None
    max_transform_queries: Optional[int] = None
    vector_top_k: Optional[int] = None
    bm25_top_k: Optional[int] = None
    rrf_mode: Optional[str] = None
    rrf_k: Optional[int] = None
    rrf_top_k: Optional[int] = None
    vector_weight: Optional[float] = None
    reranking_mode: Optional[str] = None
    rerank_top_k: Optional[int] = None
    final_top_k: Optional[int] = None
    vector_score_threshold: Optional[float] = None
    bm25_score_threshold: Optional[float] = None
    confidence_high_score_threshold: Optional[float] = None
    confidence_min_high_conf_count: Optional[int] = None
    confidence_low_score_threshold: Optional[float] = None
    extra_config: Optional[dict[str, Any]] = None


class StrategyConfigOut(StrategyConfigBase):
    """Response shape."""
    config_id: str
    kb_id: int
    version: int
    is_active: bool
    created_at: Optional[datetime] = None
    created_by: Optional[str] = None

    model_config = {"from_attributes": True}


class StrategyConfigListOut(BaseModel):
    total: int
    items: list[StrategyConfigOut]

File: management/src/api/strategy_config.ts
import { http } from "@/utils/http";
import { baseUrlApi } from "./utils";

// ──────────────────────────────────────────────
// 1. 类型定义 (对应后端 StrategyConfig 模式)
// ──────────────────────────────────────────────

export interface StrategyConfig {
  config_id: string;
  kb_id: number;
  version: number;
  is_active?: boolean;
  prompt_language?: string;
  // 组件开关
  enable_query_rewrite: boolean;
  enable_query_expansion: boolean;
  enable_query_hyde: boolean;
  enable_vector: boolean;
  enable_bm25: boolean;
  enable_reranker: boolean;
  enable_small_to_big: boolean;
  require_citation: boolean;
  // 检索参数
  query_expansion_num: number;
  max_transform_queries: number;
  vector_top_k: number;
  bm25_top_k: number;
  rrf_mode: string;
  rrf_k: number;
  rrf_top_k: number;
  vector_weight: number;
  reranking_mode: string;
  rerank_top_k: number;
  final_top_k: number;
  // 阈值参数
  vector_score_threshold: number;
  bm25_score_threshold: number;
  confidence_high_score_threshold: number;
  confidence_min_high_conf_count: number;
  confidence_low_score_threshold: number;
  // 其他
  extra_config?: Record<string, any>;
  created_at?: string;
  created_by?: string;
}

export interface KnowledgeBaseOption {
  id: number;
  name: string;
}

export type StrategyConfigCreate = StrategyConfig;
export type StrategyConfigUpdate = Partial<StrategyConfig>;

// ──────────────────────────────────────────────
// 2. API 接口定义
// ──────────────────────────────────────────────

/** 获取所有知识库选项 */
export const getKnowledgeBaseOptions = () => {
  return http.request<KnowledgeBaseOption[]>(
    "get",
    baseUrlApi("admin/knowledge-bases/options")
  );
};

/** 获取知识库下的策略列表 (带分页) */
export const getStrategyList = (
  kb_id?: number | null,
  params?: { page?: number; page_size?: number }
) => {
  return http.request<{
    total: number;
    items: StrategyConfig[];
  }>("get", baseUrlApi("admin/strategy-configs"), {
    params: {
      kb_id: kb_id ?? undefined,
      ...params
    }
  });
};

/** 获取单个策略详情 */
export const getStrategyDetail = (config_id: string) => {
  return http.request("get", baseUrlApi(`admin/strategy-configs/${config_id}`));
};

/** 获取知识库当前的激活策略 */
export const getActiveStrategy = (kb_id: number) => {
  return http.request(
    "get",
    baseUrlApi(`admin/strategy-configs/kb/${kb_id}/active`)
  );
};

/** 创建新策略 */
export const createStrategy = (data: StrategyConfigCreate) => {
  return http.request("post", baseUrlApi("admin/strategy-configs"), { data });
};

/** 部分更新策略 */
export const updateStrategy = (
  config_id: string,
  data: StrategyConfigUpdate
) => {
  return http.request(
    "patch",
    baseUrlApi(`admin/strategy-configs/${config_id}`),
    {
      data
    }
  );
};

/** 激活策略 (会自动停用该 KB 下的其他策略) */
export const activateStrategy = (config_id: string) => {
  return http.request(
    "post",
    baseUrlApi(`admin/strategy-configs/${config_id}/activate`)
  );
};

/** 删除策略 */
export const deleteStrategy = (config_id: string) => {
  return http.request(
    "delete",
    baseUrlApi(`admin/strategy-configs/${config_id}`)
  );
};
